src/heatmap.ts

```diff
@@ -0,0 +1,761 @@
+    }
+
+    this.size = size;
+    return this;
+  }
+
+  /**
+   * Set the intensity factor
+   * @param intensity - Accepts float value
+   * @returns HeatmapRenderer instance
+   */
+  setIntensity(intensity: number): HeatmapRenderer {
+    if (isNullUndefined(intensity) || isNotNumber(intensity)) {
+      this.intensity = 1.0; // applying default intensity
+      throw new Error("Invalid Intensity: Expected Number");
+    }
+
+    if (intensity > 1 || intensity < 0) {
+      this.intensity = intensity > 1 ? 1 : 0; // Setting bound value
+      throw new Error("Invalid Intensity value " + intensity);
+    }
+    this.intensity = intensity;
+    return this;
+  }
+
+  /**
+   * Set the opacity factor
+   * @param opacity - The opacity factor.
+   * @returns HeatmapRenderer instance
+   */
+  setOpacity(opacity: number): HeatmapRenderer {
+    if (isNullUndefined(opacity) || isNotNumber(opacity)) {
+      throw new Error("Invalid Opacity: Expected Number");
+    }
+
+    if (opacity > 1 || opacity < 0) {
+      throw new Error("Invalid Opacity value " + opacity);
+    }
+    this.opacity = opacity;
+    return this;
+  }
+
+  /**
+   * Set the background image
+   * @param config - Accepts Object with { Url, height, width, x, and y} properties
+   * @returns HeatmapRenderer instance
+   */
+  setBackgroundImage(config: BackgroundImageConfig) {
+    const self = this;
+    if (!config.url) {
+      return;
+    }
+
+    const maxTextureSize = this.ctx!.getParameter(this.ctx!.MAX_TEXTURE_SIZE);
+    this.imageTexture = this.ctx!.createTexture();
+    this.type = "TEXTURE_2D";
+    this.imageConfig = null;
+
+    this.imgWidth = config.width || this.width;
+    this.imgHeight = config.height || this.height;
+
+    this.imgWidth =
+      this.imgWidth > maxTextureSize ? maxTextureSize : this.imgWidth;
+    this.imgHeight =
+      this.imgHeight > maxTextureSize ? maxTextureSize : this.imgHeight;
+
+    imageInstance(
+      config.url,
+      function onUpdateCallBack(this: any) {
+        self.ctx!.activeTexture(self.ctx!.TEXTURE0);
+        self.ctx!.bindTexture(self.ctx!.TEXTURE_2D, self.imageTexture);
+        self.ctx!.texParameteri(
+          self.ctx!.TEXTURE_2D,
+          self.ctx!.TEXTURE_WRAP_S,
+          self.ctx!.CLAMP_TO_EDGE
+        );
+        self.ctx!.texParameteri(
+          self.ctx!.TEXTURE_2D,
+          self.ctx!.TEXTURE_WRAP_T,
+          self.ctx!.CLAMP_TO_EDGE
+        );
+        self.ctx!.texParameteri(
+          self.ctx!.TEXTURE_2D,
+          self.ctx!.TEXTURE_MIN_FILTER,
+          self.ctx!.LINEAR
+        );
+        self.ctx!.texParameteri(
+          self.ctx!.TEXTURE_2D,
+          self.ctx!.TEXTURE_MAG_FILTER,
+          self.ctx!.LINEAR
+        );
+
+        self.ctx!.texImage2D(
+          self.ctx!.TEXTURE_2D,
+          0,
+          self.ctx!.RGBA,
+          this.naturalWidth,
+          this.naturalHeight,
+          0,
+          self.ctx!.RGBA,
+          self.ctx!.UNSIGNED_BYTE,
+          this
```

Comment:
Wanted a confirmation whether the `this` here refers to the image element which gets created in imageInstance function.

Reply:
yes, this callback method gets invoked on Image load. "this" refers to the image context.